### ISSUE TITLE
Inconsistent PK/FK definitions

### ISSUE THREAD

**Issue by bjornharrtell** [Kortforsyningen/FIRE-DDL#17]
It seems the schema now has a mix of FK's using the integer OBJECTID in some cases and in other cases "GUID" FK. Examples of OBJECTID FK's:

https://github.com/Kortforsyningen/FIRE-DDL/blob/eaa66de58d4e7a342e1e6677ab5e2eb17f33a172/DDL.sql#L181

https://github.com/Kortforsyningen/FIRE-DDL/blob/eaa66de58d4e7a342e1e6677ab5e2eb17f33a172/DDL.sql#L18

I see no reason for this inconsistency and it's bound to cause issues and/or confuse.

I would *strongly* suggest to reevaluate using GUIDs as FKs and instead use OBJECTID FK's throughout and have GUIDs only where needed for distributed/external identification.

**Comment by bjornharrtell** [comment]
We even have a third FK type, clear text VARCHAR2(4000), as discovered in https://github.com/Kortforsyningen/FIRE-DDL/issues/16. I remain convinced it would be a good idea to OBJECTID FKs.

**Comment by busstoptaktik** [comment]
@PeterWestNielsen - I think you commented on something somewhat related to this yesterday, but I'm certain I do not understand neither the cure, nor its implications.

So would you mind commenting briefly here, so we can consider whether #17 should be closed or discussed more in depth?

**Comment by B027987** [comment]
The reason for using XYZ-OBJECTID as foreign keys is that we to point to the exact row/version.
This is a common modellling rule used througout all registers that SDFE is designing.

in the case of SAGSEVENTINFOOBJECTID  we **COULD** use ID instead due to the fact that SAGSEVENT does not contain versioning, But we prefer  to use the wellknown pattern of  SAGSEVENTINFOOBJECTID

**Comment by bjornharrtell** [comment]
@B027987 sorry but that statement confuses me. The "XYZ-OBJECTID as foreign keys" *is* what I'd like to see consistently throughout the schema. My issue is when other types of keys are used as FKs (GUID and VARCHAR2).

**Comment by busstoptaktik** [comment]
@bjornharrtell, @PeterWestNielsen & @B027987: I'm horribly confused:

1. Can we actually make do with OBJECTIDs as foreign keys, and use alphanumeric identifiers (guid, varchar) for external reference only?

2. I suppose OBJECTID is preferable for efficiency reasons, but are there other reasons

3. What are the counterarguments?

Since we are by convention forced to have an OBJECTID (aka ROWID) field on each table it seems logical to actually use it for database-internal cross reference. So if it is feasible and there are good reasons not to use guid/varchar in a FK context, I am all for restricting FKs to OBJECTIDs.

There may be modelling conventions at play here as well, since the identification of the object member OBJECTID with the table row id is implementation specific (it may change when implementing the data model in a future hypothetical **hypercubic quasiquantomatic exascalar** (TM) data storage/retrieval system).

But since, even in the data model, OBJECTID is defined as numerical, unique and monotonic, all reasoning done using the identification of OBJECTID and ROWID should be compatible with any actual representation of the data model.

**Comment by bjornharrtell** [comment]
@busstoptaktik at least your comment makes me less confused and - 

1. Yes I see no reason why not. 
2. I'm not primarily concerned with efficiency that's a nice side effect - my main concern is consistency.
3. I have none. :)

**Comment by busstoptaktik** [comment]
@bjornharrtell: I have been looking into the migration code from @nikolajkamstrup to get an idea of what kind of changes would be needed to switch to using OBJECTIDs for DB-internal reference.

My SQL-fu is, however, not sufficient to give me any feeling for whether switching would be a *piece of cake* or an *inescapable swamp of agony and despair*.

Hence, my immediate gut feeling is that this should be deferred to the next major redesign (5 years down the road).

But my gut feeling may be wrong, so @kbevers, @B027987, @PeterWestNielsen, @nikolajkamstrup: Is switching to OBJECTIDs as internal identifiers really a piece of cake and is it a good idea to do so? - please opine.

**Comment by busstoptaktik** [comment]
@bjornharrtell: I have been looking into the migration code from @nikolajkamstrup to get an idea of what kind of changes would be needed to switch to using OBJECTIDs for DB-internal reference.

My SQL-fu is, however, not sufficient to give me any feeling for whether switching would be a *piece of cake* or an *inescapable swamp of agony and despair*.

Hence, my immediate gut feeling is that this should be deferred to the next major redesign (5 years down the road).

But my gut feeling may be wrong, so @kbevers, @B027987, @PeterWestNielsen, @nikolajkamstrup: Is switching to OBJECTIDs as internal identifiers really a piece of cake and is it a good idea to do so? - please opine.

**Comment by busstoptaktik** [comment]
@bjornharrtell when you recommend using the integer OBJECTIDs, I suppose that implementing the required changes in [fireapi](https://github.com/Septima/fikspunktsregister) will be a minor task?

**Comment by bjornharrtell** [comment]
Yes, should not be a major task to adapt fireapi. Do you agree @AsgerPetersen?

**Comment by AsgerPetersen** [comment]
@bjornharrtell @busstoptaktik I think adapting fireapi will take a couple of hours (assuming the DB model is known).

**Comment by busstoptaktik** [comment]
We’ve carried out a data model review this morning, and are currently working on eliminating the varchar FKs in SRID, SRDInamespace, obstype, obstypenamespace, sagseventID.

The remaining cases of non-integral foreign keys (not to be confused with non-integral polytheism, which is outside the scope of this treatise) occur due to the invocation of corporate modelling patterns targeting a more general situation.

We have been informed from infinitely trustworthy higher data modelling deities, that these make sense in the general perspective of “the set of all SDFE data models”, and hence we defer infinitely all further work on eliminating these cases
